pkg/kubelet/util/pluginwatcher/plugin_watcher.go

```diff
@@ -26,69 +26,40 @@ import (
 	"time"
 
 	"github.com/fsnotify/fsnotify"
-	"github.com/pkg/errors"
 	"golang.org/x/net/context"
 	"google.golang.org/grpc"
 	"k8s.io/klog"
 
 	registerapi "k8s.io/kubernetes/pkg/kubelet/apis/pluginregistration/v1"
+	"k8s.io/kubernetes/pkg/kubelet/pluginmanager/cache"
 	utilfs "k8s.io/kubernetes/pkg/util/filesystem"
 )
 
 // Watcher is the plugin watcher
 type Watcher struct {
-	path           string
-	deprecatedPath string
-	stopCh         chan interface{}
-	fs             utilfs.Filesystem
-	fsWatcher      *fsnotify.Watcher
-	wg             sync.WaitGroup
-
-	mutex       sync.Mutex
-	handlers    map[string]PluginHandler
-	plugins     map[string]pathInfo
-	pluginsPool map[string]map[string]*sync.Mutex // map[pluginType][pluginName]
-}
-
-type pathInfo struct {
-	pluginType string
-	pluginName string
+	path                string
+	deprecatedPath      string
+	fs                  utilfs.Filesystem
+	fsWatcher           *fsnotify.Watcher
+	wg                  sync.WaitGroup
+	desiredStateOfWorld cache.DesiredStateOfWorld
 }
 
 // NewWatcher provides a new watcher
 // deprecatedSockDir refers to a pre-GA directory that was used by older plugins
 // for socket registration. New plugins should not use this directory.
-func NewWatcher(sockDir string, deprecatedSockDir string) *Watcher {
+func NewWatcher(sockDir string, deprecatedSockDir string, desiredStateOfWorld cache.DesiredStateOfWorld) *Watcher {
```

Comment:
nit: It would be really nice if `NewWatcher(...)` returned a Watcher *interface* instead of a struct (that would make it easier to test). If you have time, and want to, you can make this change.

Reply:
>  I plan to address with followup PRs:
>    * having NewWatcher(...) return a Watcher interface

SGTM. Could you open up a P2 cleanup issue to track this work?

Reply:
Sure. Just created #76539